src/reference_implementations/prompting_vector_llms/activation_fine_tuning/train_on_activations.ipynb

```diff
@@ -289,66 +362,135 @@
+     ]
+    }
+   ],
+   "source": [
+    "layer_numbers = [\"10\", \"20\", \"30\", \"39\"]\n",
+    "test_accuracies_by_layer = {layer_number: 0.0 for layer_number in layer_numbers}\n",
+    "device = \"cuda\" if cuda.is_available() else \"cpu\"\n",
     "\n",
-    "        logits = model(activations)\n",
-    "        loss = loss_fn(logits, labels)\n",
+    "for layer_number_to_load in layer_numbers:\n",
+    "    # Define new model\n",
+    "    model = MLP({\"embedding_dim\": 5120, \"hidden_dim\": 128, \"label_dim\": 2})\n",
+    "    device = \"cuda\" if cuda.is_available() else \"cpu\"\n",
     "\n",
-    "        loss.backward()\n",
-    "        optimizer.step()\n",
+    "    # Load the proper dataset\n",
+    "    train_dataset = ActivationDataset(\n",
+    "        os.path.join(activations_path, f\"train_activations_with_prompts_demo_{layer_number_to_load}.pkl\")\n",
+    "    )\n",
+    "    test_dataset = ActivationDataset(\n",
+    "        os.path.join(activations_path, f\"test_activations_with_prompts_demo_{layer_number_to_load}.pkl\")\n",
+    "    )\n",
+    "    train_dataloader = DataLoader(train_dataset, batch_size=128, shuffle=True, collate_fn=batch_last_token)\n",
+    "    test_dataloader = DataLoader(test_dataset, batch_size=128, shuffle=True, collate_fn=batch_last_token)\n",
     "\n",
-    "        training_params[\"Train-Loss\"] = loss.detach().item()\n",
-    "        pbar.set_postfix(training_params)\n",
+    "    # Train and evaluate\n",
+    "    test_accuracies_by_layer[layer_number_to_load] = train_and_evaluate_model(\n",
+    "        model, train_dataloader, test_dataloader, device\n",
+    "    )\n",
     "\n",
-    "    model.eval()\n",
-    "    with torch.no_grad():\n",
-    "        predictions = []\n",
-    "        for batch in test_dataloader:\n",
-    "            activations, labels = batch\n",
-    "            activations = activations.float().to(device)\n",
-    "            labels = torch.tensor(labels).to(device)\n",
+    "for layer_number in layer_numbers:\n",
+    "    print(f\"Accuracy for Layer {layer_number} WITH PROMPTS: {test_accuracies_by_layer[layer_number]}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "There are a few interesting takeaways from these results:\n",
     "\n",
-    "            logits = model(activations)\n",
-    "            predictions.extend((logits.argmax(dim=1) == labels))\n",
+    "* The prompt structures we provided helped to improve the results. For the best performing layer activations we move from `0.856%` accuracy without prompts to `0.893%` with them.\n",
     "\n",
-    "        accuracy = torch.stack(predictions).sum() / len(predictions)\n",
+    "* In both cases, the best layer activations for the task are not the first or last layers, but rather Layer 20. This has been observed in other contexts, as the earlier layers contain more general embeddings.\n",
```

Comment:
Is this also true for OPT? Namely, is the best performing layer also neither the first nor the last layer?